### ISSUE TITLE
Enable parallel testings in Unit and Integration tests and Specs

### ISSUE THREAD

**Issue by blairconrad** [FakeItEasy/FakeItEasy#774]
The Specs have had parallelization disabled since we moved to xBehave.net (running on top of xUnit.net 2), and now that we moved to xUnit.net in #545, we've had to disable parallel test runs for the integration and unit tests.
Let's see if we can remove the restriction.

**Comment by blairconrad** [comment]
Oh, that is good news about Specs and Integration, @thomaslevesque. Although I thought I saw failures in Integration. I had tried each project in turn before adding the XunitAssemblyInfo. I'll try later.
And yes, the `ConfigurableServiceLocatorTestBase` is the big scary thing in Unit. My first approach would be to see what's covered by (or can be covered by) the Specs and start culling.

**Comment by blairconrad** [comment]
It took a few attempts, but I got Specs to fail.

In `NonGenericCreationSpecs.UseSuccessfulConstructor`.

```c#
"And the one-parameter constructor was not tried"
    .x(() => FakedClass.ParameterListLengthsForAttemptedConstructors.Should().BeEquivalentTo(0, 2));
```

```
System.InvalidOperationException

Collection was modified after the enumerator was instantiated.

   at System.ThrowHelper.ThrowInvalidOperationException(ExceptionResource resource)
   at System.Collections.Generic.SortedSet`1.Enumerator.MoveNext()
   at System.Linq.Enumerable.<CastIterator>d__94`1.MoveNext()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at FluentAssertions.Collections.CollectionAssertions`2.BeEquivalentTo[T](IEnumerable`1 expected, String because, Object[] becauseArgs) in C:\projects\fluentassertions-vf06b\Src\Core\Collections\CollectionAssertions.cs:line 335
   at FakeItEasy.Specs.CreationSpecsBase.<>c.<UseSuccessfulConstructor>b__2_6() in C:\PortableApps\Documents\Source\GitHub\FakeItEasy\tests\FakeItEasy.Specs\CreationSpecs.cs:line 74
   at Xbehave.StringExtensions.<>c__DisplayClass2.<x>b__1(IStepContext c) in z:\Builds\work\f36ef1a5a4bfc851\src\Xbehave.2.Core\StringExtensions.cs:line 38
   at Xbehave.Execution.StepInvoker.<>c__DisplayClassd.<<InvokeBodyAsync>b__a>d__f.MoveNext() in z:\Builds\work\f36ef1a5a4bfc851\src\Xbehave.2.Execution.desktop\StepInvoker.cs:line 75
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Xunit.Sdk.ExecutionTimer.<AggregateAsync>d__4.MoveNext() in C:\BuildAgent\work\cb37e9acf085d108\src\xunit.execution\Sdk\Frameworks\ExecutionTimer.cs:line 48
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Xunit.Sdk.ExceptionAggregator.<RunAsync>d__9.MoveNext() in C:\BuildAgent\work\cb37e9acf085d108\src\xunit.core\Sdk\ExceptionAggregator.cs:line 90
```

This is probably because the generic creation specs were running. Makes sense.
`FakedClass.ParameterListLengthsForAttemptedConstructors` is, of course, static.

**Comment by blairconrad** [comment]
If I take a branch _without_ your changes (so `master`, as I type), then IntegrationTests fail quite consistently, with this error on 

```c#
[Fact]
public void Faked_object_with_additional_attributes_should_create_a_type_with_those_attributes()
{
    // Arrange
    var types = new Type[0];
    var constructor = typeof(ForTestAttribute).GetConstructor(types);
    var list = new object[0];
    var attribute = new CustomAttributeBuilder(constructor, list);
    var builder = new List<CustomAttributeBuilder> { attribute };

    // Act
    var fake = A.Fake<IEmpty>(a => a.WithAdditionalAttributes(builder));

    // Assert
    fake.GetType().GetCustomAttributes(typeof(ForTestAttribute), true).Should().HaveCount(1);
}
```

```
FluentAssertions.Execution.AssertionFailedException

Expected collection to contain 1 item(s), but found 0.

   at FluentAssertions.Execution.FallbackTestFramework.Throw(String message) in C:\projects\fluentassertions-vf06b\Src\Shared\Execution\FallbackTestFramework.cs:line 21
   at FluentAssertions.Execution.AssertionScope.FailWith(String message, Object[] args) in C:\projects\fluentassertions-vf06b\Src\Core\Execution\AssertionScope.cs:line 197
   at FluentAssertions.Collections.SelfReferencingCollectionAssertions`2.HaveCount(Int32 expected, String because, Object[] becauseArgs) in C:\projects\fluentassertions-vf06b\Src\Core\Collections\SelfReferencingCollectionAssertions.cs:line 53
```

And indeed, I would expect your fix to solve this.

**Comment by blairconrad** [comment]
@thomaslevesque, without your change, I see a couple of extra Specs failing as well. On `NonGenericCreationOptionsSpecs`. To do with missing attributes again. You are on fire, sir.

**Comment by thomaslevesque** [comment]
> It took a few attempts, but I got Specs to fail.

With my changes, you mean? I ran them several times without seeing any failure, but that doesn't mean much... by nature, concurrency bugs happen randomly.

**Comment by blairconrad** [comment]
> > It took a few attempts, but I got Specs to fail.
>
> With my changes, you mean? I ran them several times without seeing any failure, but that doesn't mean much... by nature, concurrency bugs happen randomly.

Yup. I should've been more clear. It was worse without your changes. The problem with `NonGenericCreationSpecs.UseSuccessfulConstructor` is obvious, in retrospect, and probably workaroundable.

**Comment by adamralph** [comment]
🎉 nice work!

**Comment by blairconrad** [comment]
This issue was fixed in [Release 3.2.0](https://github.com/FakeItEasy/FakeItEasy/releases/tag/3.2.0).
